File: problems/0003_longest-substring-without-repeating-characters/map/Solution.scala
import scala.collection.mutable.Map

object Solution {
    def lengthOfLongestSubstring(s: String): Int = {
        val map = Map[Char, Int]()
        var (max, start, end) = (0, 0, 0)
        for (c <- s) {
            val x = map.get(c)
            map += (c -> end)
            end += 1
            x match {
                case Some(idx) if idx >= start =>
                    start = idx + 1
                case _ =>
                    max = (end - start).max(max)
            }
        }
        max
    }
}

File: problems/0003_longest-substring-without-repeating-characters/map/solution.ts
function lengthOfLongestSubstring(s: string): number {
    const map: { [index: string]: number } = {}
    let max = 0
    for (let start = 0, end = 0; end < s.length;) {
        const c = s[end],
              idx = map[c]
        map[c] = end++
        if (idx === undefined || idx < start)
            max = Math.max(max, end - start)
        else
            start = idx + 1
    }
    return max
};
